e2e/txpool_test.go

```diff
@@ -0,0 +1,186 @@
+}
+
+func generateReq(params generateTxReqParams) *txpoolOp.AddTxnReq {
+	msg := &txpoolOp.AddTxnReq{
+		Raw: &any.Any{
+			Value: generateTx(params).MarshalRLP(),
+		},
+		From: types.ZeroAddress.String(),
+	}
+
+	return msg
+}
+
+func TestTxpool_ErrNonceTooLow(t *testing.T) {
+	// Test scenario:
+	// Add tx with nonce 0
+	// -> Check if tx has been parsed
+	// Add tx with nonce 0
+	// -> tx shouldn't be added, since the nonce is too low
+
+	// Predefined values
+	gasPrice := big.NewInt(10000)
+
+	referenceKey, referenceAddr := tests.GenerateKeyAndAddr(t)
+	defaultBalance := framework.EthToWei(10)
+
+	devInterval := 5
+
+	// Set up the test server
+	srvs := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
+		config.SetConsensus(framework.ConsensusDev)
+		config.SetSeal(true)
+		config.SetDevInterval(devInterval)
+		config.Premine(referenceAddr, defaultBalance)
+	})
+	srv := srvs[0]
+
+	// TxPool client
+	clt := srv.TxnPoolOperator()
+	toAddress := types.StringToAddress("1")
+
+	// Add the initial transaction
+	addReq := generateReq(generateTxReqParams{
+		nonce:         0,
+		referenceAddr: referenceAddr,
+		referenceKey:  referenceKey,
+		toAddress:     toAddress,
+		gasPrice:      gasPrice,
+		value:         oneEth,
+		t:             t,
+	})
+
+	_, addErr := clt.AddTxn(context.Background(), addReq)
+	if addErr != nil {
+		t.Fatalf("Unable to add txn, %v", addErr)
+	}
+
+	// Wait for the state transition to be executed
+	_ = waitForBlock(t, srv, 1, 0)
+
+	// Add the transaction with lower nonce value than what is
+	// currently in the world state
+	_, addErr = clt.AddTxn(context.Background(), addReq)
+	assert.NotNil(t, addErr)
+	assert.Contains(t, addErr.Error(), txpool.ErrNonceTooLow.Error())
+}
+
+func TestTxpool_ErrInsufficientFunds(t *testing.T) {
+	// Test scenario:
+	// Add tx with insufficient funds
+	// -> Tx should be discarded because of low funds
+
+	// Predefined values
+	gasPrice := big.NewInt(10000)
+
+	referenceKey, referenceAddr := tests.GenerateKeyAndAddr(t)
+	defaultBalance := framework.EthToWei(1)
+
+	devInterval := 5
+
+	// Set up the test server
+	srvs := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
+		config.SetConsensus(framework.ConsensusDev)
+		config.SetSeal(true)
+		config.SetDevInterval(devInterval)
+		config.Premine(referenceAddr, defaultBalance)
+	})
+	srv := srvs[0]
+
+	// TxPool client
+	clt := srv.TxnPoolOperator()
+	toAddress := types.StringToAddress("1")
+
+	// Add the transaction
+	addReq := generateReq(generateTxReqParams{
+		nonce:         0,
+		referenceAddr: referenceAddr,
+		referenceKey:  referenceKey,
+		toAddress:     toAddress,
+		gasPrice:      gasPrice,
+		value:         framework.EthToWei(5),
+		t:             t,
+	})
+
+	// The transaction shouldn't be added because of low funds
+	_, addErr := clt.AddTxn(context.Background(), addReq)
+	assert.NotNil(t, addErr)
+	assert.Contains(t, addErr.Error(), txpool.ErrInsufficientFunds.Error())
+}
```

Comment:
GJ on these tests.
Could you maybe make one test from them with a test table?

Reply:
Done 👍 

d14bf63